--- libwebsockets-v4.3-stable/test-apps/CMakeLists.txt.orig
+++ libwebsockets-v4.3-stable/test-apps/CMakeLists.txt
@@ -88,7 +88,7 @@
 			endif()
 			target_link_libraries(${TEST_NAME} websockets)
 			add_dependencies(${TEST_NAME} websockets)
-			if (UNIX AND LWS_WITH_SSL AND NOT LWS_WITH_MBEDTLS)
+			if (UNIX AND NOT LWS_WITH_MBEDTLS)
 				target_link_libraries(${TEST_NAME} ${CMAKE_DL_LIBS})
 			endif()
 		endif()
@@ -115,7 +115,7 @@
 			PROPERTIES
 			OUTPUT_NAME libwebsockets-${TEST_NAME})
 			
-		target_link_libraries(${TEST_NAME} ${LIB_LIST_AT_END})
+		target_link_libraries(${TEST_NAME} ${LIB_LIST_AT_END} "-ldl")
 
 		# Add to the list of tests.
 		list(APPEND TEST_APP_LIST ${TEST_NAME})
